### ISSUE TITLE
False positive `E0601: used-before-assignment` when assignment of variables is inside `if __name__ == "__main__":`

### ISSUE THREAD

**Issue by shatakshiiii** [pylint-dev/pylint#8543]
### Bug description

Pylint version **3.0.0a6** seems to produce an `used-before-assignment` error when variables are assigned within 
`if __name__ == "__main__":` function and later used in different functions.

This is coming from latest pylint v3.0.0a6 (since there is no such error in previous pylint versions) 

Example code to reproduce this error:

```
def foo():
    print(x-y)

def main():
    print(x+y)
    foo()

if __name__ == "__main__":
    x = 20
    y = 10
    main()
 ```
 
Logging this issue here to get more context around this, as this is just a false-positive with **v3.0.0a6** or something that pylint is aiming to trigger in its newer versions (As the above code format is supported in python).

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:2:10: E0601: Using variable 'x' before assignment (used-before-assignment)
example.py:2:12: E0601: Using variable 'y' before assignment (used-before-assignment)
```

### Expected behavior

Pylint should not trigger `used-before-asssignment` error.

### Pylint version

```shell
pylint 3.0.0a6
astroid 2.15.2
Python 3.9.0 (default, Oct  6 2022, 15:58:26)
```
```


### OS / Environment

Fedora 37

### Additional dependencies

_No response_

**Comment by jacobtylerwalls** [comment]
Thanks for the question. This is intentional as of #6677. You're correct that this executes just fine when run from the CLI, but any other uses will raise NameError:

```python
>>> import a
>>> a.foo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jacobwalls/pylint/a.py", line 2, in foo
    print(x-y)
          ^
NameError: name 'x' is not defined
```

**Comment by shatakshiiii** [comment]
Thanks for the explaination @jacobtylerwalls !
